In [1]:
import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
import plotly.express as px
from plotly.subplots import make_subplots
import plotly.io as pio
pio.renderers.defaule = 'colab'

from itables import show

# This stops a few warning messages from showing
pd.options.mode.chained_assignment = None 
import warnings
warnings.simplefilter(action='ignore', category=FutureWarning)

In [3]:
file_name = 'data/yougov-climate.csv'
DF = pd.read_csv(file_name,index_col=0)
DF

The climate is changing and human activity is mainly responsible    \
country                                                                       
GB                                                     833                    
US                                                     507                    

         The climate is changing and human activity is partly responsible, together with other factors    \
country                                                                                                    
GB                                                     604                                                 
US                                                     493                                                 

         The climate is changing but human activity is not responsible at all    \
country                                                                           
GB                                                      49                        
US                                                     120                        

         The climate is not changing    Don't know  
country                                             
GB                                  33         114  
US                                  80         133

In the lecture we analyzed the following questions for the first column in the data:

### First question:

What percent of **all respondents** think the climate is changing and human activity is mainly responsible?  

### Second question:

What percent of **GB respondents** think the climate is changing and  human activity is mainly responsible?  

### Third Question

What percent of **US respondents** think the climate is changing and  human activity is mainly responsible?  

**Q1** Repeat the analysis from the lecture notes for one of the other columns/questions.

1. Percent total
2. Percent from GB
3. Percent from US

Talk about the conditional probability in this case:

eg. In a person is from ________ then there is a ________ probability that they believe _______. If a person answered _______ then they are more likley to be from ________. 

**First Question:**

In [7]:
DF.loc['total']= DF.sum(axis=0)
DF['total'] = DF.sum(axis=1)
DF

The climate is changing and human activity is mainly responsible    \
country                                                                       
GB                                                     833                    
US                                                     507                    
total                                                 1340                    

         The climate is changing and human activity is partly responsible, together with other factors    \
country                                                                                                    
GB                                                     604                                                 
US                                                     493                                                 
total                                                 1097                                                 

         The climate is changing but human activity is not responsible at all    \
country                                                                           
GB                                                      49                        
US                                                     120                        
total                                                  169                        

         The climate is not changing    Don't know  total  
country                                                    
GB                                  33         114   1633  
US                                  80         133   1333  
total                              113         247   2966

In [9]:
all_respondents = DF['total'].loc['total']
human_responsible = DF['The climate is changing and human activity is mainly responsible  ']['total']
human_responsible/all_respondents

0.45178691840863117

The first code added a new observation and a new column that added together the toal of each column and the total for each observation. The second code found the percentage of the total amount of people that voted that humans were mainly responsible for climate change versus the complete total of respondents.

**First Question Answer:** This means that overall 45.2% of people think that climate is changing and human activity is mainly responsible.

**Second Question:**

In [14]:
gb_respondants = DF['total'].loc['GB']
GB_human_responsible = DF['The climate is changing and human activity is mainly responsible  ']['GB']
GB_human_responsible/gb_respondants

0.5101041028781383

The code above exchanged the ".loc'GB'" command to focus on another observation; the total amount of GB respondents instead of the entire total. Then the code divided the total amount of British people that voted that humans were mainly responsible for climate change verus the total amount of British respondents.

**Second Question Answer:** Of the people would responded from Great Britain, 51% think that climate is changing and human activity is mainly responsible.

**Third Question:**

In [19]:
us_respondants = DF['total'].loc['US']
US_human_responsible = DF['The climate is changing and human activity is mainly responsible  ']['US']
US_human_responsible/us_respondants

0.3803450862715679

The code above exchanged the ".loc'US'" command to focus on another observation; the total amount of US respondents instead of the entire total. Then the code divided the total amount of American people that voted that humans were mainly responsible for climate change verus the total amount of American respondents.

**Third Question Answer:** When we consider US respondents we see that 38% (lower than GB) believe that climate is changing and human activity is mainly responsible.

**Q1:**
1) Percent Total of respondents that don't believe that the climate is changing.

In [36]:
All_respondents = DF['total'].loc['total']
Human_responsible = DF['The climate is not changing  ']['total']
Human_responsible/All_respondents

0.03809844908968307

The code above ran that complete total of respondents that believed that the climate is not chaning. Which was 3.8%.

**Q1:**

2) Percent from GB respondents that don't believe that the climate is changing.

In [44]:
GB_respondents = DF['total'].loc['GB']
GB_Human_responsible = DF['The climate is not changing  ']['GB']
GB_Human_responsible/GB_respondents

0.02020820575627679

The code above ran the total amount of British respondents that believe that the climate is not changing. The British percentage for this was 2%.

**Q1:**

3) Percent from US respondents that don't believe that the climate is changing.

In [46]:
US_respondents = DF['total'].loc['US']
US_Human_responsible = DF['The climate is not changing  ']['US']
US_Human_responsible/US_respondents

0.06001500375093773

The code above ran the total amount of American respondents that believe that the climate is not changing. The American percentage for this was 6%.

**Q1 Conditional Probability:** If a person is from the US then there is a 6% probability that they believe that the climate is not changing. If a person answered that the climate is changing then they are more likley to be from GB.

## Berkeley admission data example

- Study carried out by the Graduate Division of the University of California, Berkeley in the early 70’s to evaluate whether there was a gender bias in graduate admissions.
- The data come from six departments. For confidentiality we'll call them A-F. 
- We have information on whether the applicant was male or female and whether they were admitted or rejected. This is an old study so only two binary classifications were used. 

### Here is the data


In [52]:
file_name = 'data/berkley.csv'
DF = pd.read_csv(file_name)
show(DF)

In [54]:
DF_melt = pd.melt(DF,id_vars=['Department'],var_name='MF',value_name='Number')
DF_melt

Department          MF  Number
0           A    Male Yes     512
1           B    Male Yes     353
2           C    Male Yes     120
3           D    Male Yes     138
4           E    Male Yes      53
5           F    Male Yes      22
6           A     Male No     313
7           B     Male No     207
8           C     Male No     205
9           D     Male No     279
10          E     Male No     138
11          F     Male No     351
12          A  Female Yes      89
13          B  Female Yes      17
14          C  Female Yes     202
15          D  Female Yes     131
16          E  Female Yes      94
17          F  Female Yes      24
18          A   Female No      19
19          B   Female No       8
20          C   Female No     391
21          D   Female No     244
22          E   Female No     299
23          F   Female No     317

In [56]:
DF_melt['gender'] = DF_melt['MF'].apply(lambda x: x.split(' ')[0]).copy()
DF_melt['admitted'] = DF_melt['MF'].apply(lambda x: x.split(' ')[1]).copy()

In [58]:
DF_melt

Department          MF  Number  gender admitted
0           A    Male Yes     512    Male      Yes
1           B    Male Yes     353    Male      Yes
2           C    Male Yes     120    Male      Yes
3           D    Male Yes     138    Male      Yes
4           E    Male Yes      53    Male      Yes
5           F    Male Yes      22    Male      Yes
6           A     Male No     313    Male       No
7           B     Male No     207    Male       No
8           C     Male No     205    Male       No
9           D     Male No     279    Male       No
10          E     Male No     138    Male       No
11          F     Male No     351    Male       No
12          A  Female Yes      89  Female      Yes
13          B  Female Yes      17  Female      Yes
14          C  Female Yes     202  Female      Yes
15          D  Female Yes     131  Female      Yes
16          E  Female Yes      94  Female      Yes
17          F  Female Yes      24  Female      Yes
18          A   Female No      19  Female       No
19          B   Female No       8  Female       No
20          C   Female No     391  Female       No
21          D   Female No     244  Female       No
22          E   Female No     299  Female       No
23          F   Female No     317  Female       No

### Gender distribution by department

What can we say about the gender distribution if we look at the individual departments.

Start with our original "melted" data frame:

#### Lets pivot!

Pivot this data so that our departments become the column labels and our MF column becomes the index.

In [70]:
# Your code here
DF_dept = DF_melt.pivot(index='MF',columns='Department',values='Number')
show(DF_dept)

**Q2** Calculate the proportions for the other departments. We did department A in the lecture. You can use the code from the lecture or come up with your own way to do this. You can do this one by one or try using a FOR loop.

Talk about what you see in these proportions. Which departments admit a large number of women? How do your number align with the visualization below?

##### Proportion for Dep. A: Women=82.4% & Male=62.1%

In [87]:
# Your code here
prob_female_A = DF_dept['A']['Female Yes']/(DF_dept['A']['Female Yes']+DF_dept['A']['Female No'])
prob_female_A

0.8240740740740741

In [89]:
prob_male_A = DF_dept['A']['Male Yes']/(DF_dept['A']['Male Yes']+DF_dept['A']['Male No'])
prob_male_A

0.6206060606060606

##### Proportion for Dep. B: Women=68% & Male=63%

In [91]:
prob_female_B = DF_dept['B']['Female Yes']/(DF_dept['B']['Female Yes']+DF_dept['B']['Female No'])
prob_female_B

0.68

In [95]:
prob_male_B = DF_dept['B']['Male Yes']/(DF_dept['B']['Male Yes']+DF_dept['B']['Male No'])
prob_male_B

0.6303571428571428

##### Proportion for Dep. C: Women=34.1% & Male=36.9%

In [100]:
prob_female_C = DF_dept['C']['Female Yes']/(DF_dept['C']['Female Yes']+DF_dept['C']['Female No'])
prob_female_C

0.3406408094435076

In [102]:
prob_male_C = DF_dept['C']['Male Yes']/(DF_dept['C']['Male Yes']+DF_dept['C']['Male No'])
prob_male_C

0.36923076923076925

##### Proportion for Dep. D: Women=34.9% & Male=33.1%

In [105]:
prob_female_D = DF_dept['D']['Female Yes']/(DF_dept['D']['Female Yes']+DF_dept['D']['Female No'])
prob_female_D

0.34933333333333333

In [107]:
prob_male_D = DF_dept['D']['Male Yes']/(DF_dept['D']['Male Yes']+DF_dept['D']['Male No'])
prob_male_D

0.33093525179856115

##### Proportion for Dep. E: Women=23.9% & Male=27.7%

In [110]:
prob_female_E = DF_dept['E']['Female Yes']/(DF_dept['E']['Female Yes']+DF_dept['E']['Female No'])
prob_female_E

0.23918575063613232

In [112]:
prob_male_E = DF_dept['E']['Male Yes']/(DF_dept['E']['Male Yes']+DF_dept['E']['Male No'])
prob_male_E

0.2774869109947644

##### Proportion for Dep. F: Women=7% & Male=5.9%

In [115]:
prob_female_F = DF_dept['F']['Female Yes']/(DF_dept['F']['Female Yes']+DF_dept['F']['Female No'])
prob_female_F

0.07038123167155426

In [117]:
prob_male_F = DF_dept['F']['Male Yes']/(DF_dept['F']['Male Yes']+DF_dept['F']['Male No'])
prob_male_F

0.058981233243967826

**Q2 Answers:** I noticed that 4 out of the 7 departments had a higher percentage of women in comparison to men. These 4 departments were Department A, Department B, Department D, and Department F. Additionally, my numbers align with the visualization below perfectly because all of the percentages for each of the departments "line-up". For example, the 4 departments that had a higher percentage of women are displayed in the graph, even the amount by which the percentage leads. And the percentages of women that were lower in certain departments are also shown on the graphs.

### Here is a plot of the proportions data

In [23]:
fig = px.histogram(DF_melt,
                   y='gender',
                   x='Number',
                   barnorm = "percent",
                   color='admitted',
                   facet_col='Department',
                   facet_col_wrap=2,
                   color_discrete_map = {'No':'#99d8c9','Yes':'#2ca25f'})

fig.for_each_annotation(lambda a: a.update(text=a.text.split("=")[1]))
fig.update_xaxes(title_text='')

fig.update_layout(title='Percent male and female applications',
                  title_x=0.5,
                  template="plotly_white",
                  xaxis_title="",
                  yaxis_title="Gender",
                  legend_title='Admitted',
                  autosize=False,
                  width=800,
                  height=500)

fig.show()

**Simpson's Paradox:** When we see a trend when we look at the data grouping one way, but the trend disappears or reverses when we look at the data another way. And, Simpson’s paradox illustrates the effect that omission of an explanatory variable can have on the measure of association between another explanatory variable and a response variable.

In [135]:
DF['totals'] = DF[['Male Yes', 'Male No', 'Female Yes', 'Female No']].sum(axis=1)

In [137]:
DF_male = DF[['Department', 'Male Yes', 'Male No']].copy()
DF_female = DF[['Department', 'Female Yes', 'Female No']].copy()

In [141]:
DF_male['Yes'] = DF_male['Male Yes']/(DF_male['Male Yes']+DF_male['Male No'])
DF_male['No'] = DF_male['Male No']/(DF_male['Male Yes']+DF_male['Male No'])
DF_male['proportion'] = (DF_male['Male Yes']+DF_male['Male No'])/DF['totals']
DF_male

Department  Male Yes  Male No       Yes        No  proportion
0          A       512      313  0.620606  0.379394    0.884244
1          B       353      207  0.630357  0.369643    0.957265
2          C       120      205  0.369231  0.630769    0.354031
3          D       138      279  0.330935  0.669065    0.526515
4          E        53      138  0.277487  0.722513    0.327055
5          F        22      351  0.058981  0.941019    0.522409

In [143]:
DF_female['Yes'] = DF_female['Female Yes']/(DF_female['Female Yes']+DF_female['Female No'])
DF_female['No'] = DF_female['Female No']/(DF_female['Female Yes']+DF_female['Female No'])
DF_female['proportion'] = (DF_female['Female Yes']+DF_female['Female No'])/DF['totals']
DF_female

Department  Female Yes  Female No       Yes        No  proportion
0          A          89         19  0.824074  0.175926    0.115756
1          B          17          8  0.680000  0.320000    0.042735
2          C         202        391  0.340641  0.659359    0.645969
3          D         131        244  0.349333  0.650667    0.473485
4          E          94        299  0.239186  0.760814    0.672945
5          F          24        317  0.070381  0.929619    0.477591

##### Examples:
##### In Department A, 88.4% applications come from males and only 11.6% from females - this department admitted 512 males and 89 females. So even thought the acceptance rate for females was 82.4% this department still has a small overall number of females.
##### In the smaller Department F, they admitted about the same numbers of males and females and they got about the same number of applications from each gender.

------------------------
------------------------

## Homework - Exercises

** Homework content and data from Data Science in a Box - he-05-legos.Rmd

This week we'll do some data gymnastics to refresh and review what we learned over the past few weeks using (**simulated**) data from Lego sales in 2018 for a sample of customers who bought Legos in the US. This is different than the data we used on the Exam!


### Data

In [149]:
file_name = 'data/lego-sales.csv'
DF = pd.read_csv(file_name)
DF

first_name  last_name  age  phone_number  set_id   number  \
0     Kimberly  Beckstead   24  216-555-2549   24701    76062   
1         Neel     Garvin   35  819-555-3189   25626    70595   
2         Neel     Garvin   35  819-555-3189   24665    21031   
3      Chelsea   Bouchard   41           NaN   24695    31048   
4      Chelsea   Bouchard   41           NaN   25626    70595   
..         ...        ...  ...           ...     ...      ...   
615     Talise   Nieukirk   16  801-555-2343   24902    41556   
616    Spencer     Morgan   28  784-555-3455   26041    41580   
617    Spencer     Morgan   28  784-555-3455   26060  5005051   
618     Amelia    Hageman   40  336-555-1950   24702    76063   
619     Amelia    Hageman   40  336-555-1950   24720    10830   

                      theme              subtheme  year  \
0    DC Comics Super Heroes         Mighty Micros  2018   
1                   Ninjago  Rise of the Villains  2018   
2              Architecture                   NaN  2018   
3                   Creator                   NaN  2018   
4                   Ninjago  Rise of the Villains  2018   
..                      ...                   ...   ...   
615                  Mixels              Series 7  2018   
616                  Mixels              Series 9  2018   
617                    Gear         Digital Media  2018   
618  DC Comics Super Heroes         Mighty Micros  2018   
619                   Duplo                   NaN  2018   

                                       name  pieces  us_price  \
0                            Robin vs. Bane    77.0      9.99   
1                      Ultra Stealth Raider  1093.0    119.99   
2                              Burj Khalifa   333.0     39.99   
3                            Lakeside Lodge   368.0     29.99   
4                      Ultra Stealth Raider  1093.0    119.99   
..                                      ...     ...       ...   
615                                  Tiketz    62.0      4.99   
616                                    Myke    63.0      4.99   
617  Friends of Heartlake City Girlz 4 Life     NaN     19.99   
618              The Flash vs. Captain Cold    88.0      9.99   
619                           Minnie's Café    27.0     19.99   

                                             image_url  quantity  
0    http://images.brickset.com/sets/images/76062-1...         1  
1    http://images.brickset.com/sets/images/70595-1...         1  
2    http://images.brickset.com/sets/images/21031-1...         1  
3    http://images.brickset.com/sets/images/31048-1...         1  
4    http://images.brickset.com/sets/images/70595-1...         1  
..                                                 ...       ...  
615  http://images.brickset.com/sets/images/41556-1...         2  
616                                                NaN         2  
617                                                NaN         1  
618  http://images.brickset.com/sets/images/76063-1...         2  
619  http://images.brickset.com/sets/images/10830-1...         4  

[620 rows x 14 columns]

### Exercises

* Answer the following questions using reproducible Python code.
* For each question, state your answer in a sentence, e.g. "In this sample, the first three common names of purchasers are ...".
* Note that the answers to all questions are within the context of this particular sample of sales, i.e. you shouldn't make inferences about the population of all Lego sales based on this sample.

-------------

1.  Describe what you see in the data set (variables, observations, etc)

2.  What are the three most common first names of purchasers?

3.  What are the three most common themes of Lego sets purchased?

4.  Among the most common theme of Lego sets purchased, what is the most common subtheme?

5.  Create data frames for each of the ages in the following categories: "18 and under", "19 - 25", "26 - 35", "36 - 50", "51 and over". HINT - use masks and create separate data frames -OR- create a new column with these categorical labels (more advanced).

6.  Which age group has purchased the highest number of Lego sets.

7.  Which age group has spent the most money on Legos?

8.  Which Lego theme has made the most money for Lego? HINT: Simpler than #5, just use a groupby()

9.  Which area code has spent the most money on Legos? In the US the area code is the first 3 digits of a phone number. HINT: You will need to split the phone number and get just the first three. You decided what to do about reporting the NaNs.

10.  Come up with a question you want to answer using these data, and write it down. Then, create a data visualization that answers the question, and explain how your visualization answers the question.

## Essay Response

Give a brief report on what you are reading and what you have learned so far. If you chose your own articles please insert a link to the articles. If you chose a book, please give the book title and author.

1. What drew you to the book/articles
2. What have you learned so far.

eg. In the first chapter of the book the author talks about..... or My first article discusses..... or I disagree with what the author is saying so far.... or I really like what I am learning about because....